src/python/pants/build_graph/build_configuration.py

```diff
@@ -3,30 +3,81 @@
 
 import logging
 import typing
+from collections import defaultdict
 from collections.abc import Iterable
 from dataclasses import dataclass, field
-from typing import Any, Dict, Type, Union
+from enum import Enum
+from typing import Any, Dict, Mapping, Set, Type, Union, cast
 
 from pants.build_graph.build_file_aliases import BuildFileAliases
+from pants.engine.goal import GoalSubsystem
 from pants.engine.rules import Rule, RuleIndex
 from pants.engine.target import Target
 from pants.engine.unions import UnionRule
+from pants.goal.run_tracker import RunTracker
+from pants.option.global_options import GlobalOptions
 from pants.option.optionable import Optionable
+from pants.reporting.reporting import Reporting
 from pants.util.ordered_set import FrozenOrderedSet, OrderedSet
+from pants.vcs.changed import Changed
 
 logger = logging.getLogger(__name__)
 
 
+# No goal or target_type can have a name from this set, so that `./pants help <name>`
+# is unambiguous.
+_RESERVED_NAMES = {"global", "targets", "goals"}
+
+
+# Subsystems used outside of any rule.
+_GLOBAL_SUBSYSTEMS: FrozenOrderedSet[Type[Optionable]] = FrozenOrderedSet(
+    {GlobalOptions, Reporting, RunTracker, Changed}
+)
+
+
 @dataclass(frozen=True)
 class BuildConfiguration:
     """Stores the types and helper functions exposed to BUILD files."""
 
     registered_aliases: BuildFileAliases
-    optionables: FrozenOrderedSet[Optionable]
+    optionables: FrozenOrderedSet[Type[Optionable]]
     rules: FrozenOrderedSet[Rule]
     union_rules: FrozenOrderedSet[UnionRule]
     target_types: FrozenOrderedSet[Type[Target]]
 
+    @property
+    def all_optionables(self) -> FrozenOrderedSet[Type[Optionable]]:
+        """Return all optionables in the system: global and those registered via rule usage."""
+        return _GLOBAL_SUBSYSTEMS | self.optionables
+
+    def __post_init__(self) -> None:
```

Comment:
I don't think this will catch collisions between the subsystem `pex-binary` and the target type `pex_binary`. As discussed, that's subtle and I think we do want to error here.

To fix, I think for subsystems you can add both `my-scope` and `my_scope` to the defaultdict? We know target types can only use `_` due to Python restrictions.

Reply:
Was going to do that in a followup, along with actually allowing the normalized form in config files, but yeah I can add just the check here now.